api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,233 @@
+        if(context.ctx_map_ == ctx_map_){
+          return true;
+        }
+        else{
+          return false;
+        }
+      }
+
+      /* Returns the value associated with the passed in key */
+      int GetValue(ContextKey key){
+        return ctx_map_[key.GetIdentifier()];
+      } 
+
+      /* Returns a ContextKey that has the passed in name and the
+       * next available identifier.*/
+      ContextKey CreateKey(std::string key_name){
+        int id;
+        
+        context_id_mutex.lock();
+        
+        last_key_identifier_++;
+        
+        id = last_key_identifier_;
+        
+        context_id_mutex.unlock();
+            
+        return ContextKey(key_name,id); 
+      } 
+
+    private:
+
+      /* The identifier itself */
+      std::map<int, int> ctx_map_;
+
+      /*Used to track that last ContextKey identifier and create the next one */
+      static int last_key_identifier_;
+      
+      /* A constructor that accepts a key/value map */
+      Context(std::map<int,int> ctx_map){
+        ctx_map_ = ctx_map;
+      } 
+
+
+  };
+
+
+
+  /* The token class provides an identifier that is used by
+   * the attach and detach methods to keep track of context 
+   * objects. 
+   */
+
+  class Token{
+
+    public:
+      
+      /* A constructor that sets the token's Context object to the
+       * one that was passed in. 
+       */
+      Token(Context &ctx){
+        ctx_ = ctx;  
+      } 
+
+      /* Returns the stored context object */
+      Context GetContext(){
+        return ctx_;
+      }
+
+    private:
+
+      Context ctx_;
+  };
+
+
+  /* The RuntimeContext class provides a wrapper for 
+   * propogating context through cpp. */
+  class RuntimeContext {
+
+    public:
+
+
+      /* A default constructor that will set the context to
+       * an empty context object.
+       */
+      RuntimeContext(){
+        context_ = Context();
+      }
+
+      /* A constructor that will set the context as the passed in context. */
+      RuntimeContext(Context &context){ 
+        context_ = context;
+      }
+
+      /* Sets the current 'Context' object. Returns a token 
+       * that can be used to reset to the previous Context.
+       */
+      Token Attach(Context &context){
+
+        Token old_context_token = Token(context_);
+
+        context_ = context;
+
+        return old_context_token;
+      }
+
+
+      /* Return the current context. */
+      static Context GetCurrent(){
+        Context context = context_;
+        return context_;  
+      }
+
+
+      /* Resets the context to a previous value stored in the 
+       * passed in token. Returns zero if successful, -1 otherwise
+       */
+      int Detach(Token &token){
+        
+        if(token.GetContext() == context_){
+
+          return -1;
+        }
+        
+        context_ = token.GetContext();
+         
+        return 0;     
+      }
+
+
+    private:
+     
+      static thread_local Context context_;
```

Comment:
Not 100% sure this static would always work in certain DLL loading scenarios, where two different DLLs consume OT API surface, esp. lazy loading / delay-loading. May want to spell out the limitations somewhere in the class documentation..

https://devblogs.microsoft.com/oldnewthing/20101122-00/?p=12233

Reply:
Agreed. Rather use this [singleton style](http://laristra.github.io/flecsi/src/developer-guide/patterns/meyers_singleton.html) (adding `thread_local`). That's what we use [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/provider.h#L46).